polar-core/src/inverter.rs

```diff
@@ -0,0 +1,203 @@
+use crate::runnable::Runnable;
+use crate::terms::{Operation, Operator, Term, Value};
+use crate::vm::{Binding, BindingStack, Goals, PolarVirtualMachine};
+
+#[derive(Clone)]
+pub struct Inverter {
+    vm: PolarVirtualMachine,
+    bindings: Rc<RefCell<BindingStack>>,
+    bsp: usize,
+    results: Vec<BindingStack>,
+}
+
+impl Inverter {
+    pub fn new(
+        vm: &PolarVirtualMachine,
+        goals: Goals,
+        bindings: Rc<RefCell<BindingStack>>,
+        bsp: usize,
+    ) -> Self {
+        Self {
+            vm: vm.clone_with_bindings(goals),
+            bindings,
+            bsp,
+            results: vec![],
+        }
+    }
+}
+
+struct ConstraintInverter {
+    pub new_bindings: BindingStack,
+}
+
+impl ConstraintInverter {
+    pub fn new() -> Self {
+        Self {
+            new_bindings: vec![],
+        }
+    }
+}
+
+impl Folder for ConstraintInverter {
+    /// Invert operators.
+    fn fold_operator(&mut self, o: Operator) -> Operator {
+        match o {
+            Operator::And => Operator::Or,
+            Operator::Or => Operator::And,
+            Operator::Unify | Operator::Eq => Operator::Neq,
+            Operator::Neq => Operator::Unify,
+            Operator::Gt => Operator::Leq,
+            Operator::Geq => Operator::Lt,
+            Operator::Lt => Operator::Geq,
+            Operator::Leq => Operator::Gt,
+            _ => todo!("negate {:?}", o),
+        }
+    }
+
+    /// Invert constraints.
+    fn fold_constraints(&mut self, c: Constraints) -> Constraints {
+        let partial = match c.operations.len() {
+            // Do nothing to an empty partial.
+            0 => return c,
+
+            // Invert a single constraint.
+            1 => Constraints {
+                variable: c.variable.clone(),
+                operations: vec![fold_operation(c.operations[0].clone(), self)],
+            },
+
+            // Invert the conjunction of multiple constraints, yielding a disjunction of their
+            // inverted selves. (De Morgan's Law)
+            _ => Constraints {
+                variable: c.variable.clone(),
+                operations: vec![Operation {
+                    operator: Operator::Or,
+                    args: c
+                        .operations
+                        .into_iter()
+                        .map(|o| Term::new_temporary(Value::Expression(fold_operation(o, self))))
+                        .collect(),
+                }],
+            },
+        };
+        self.new_bindings.push(Binding(
+            partial.variable.clone(),
+            Term::new_temporary(Value::Partial(partial.clone())),
+        ));
+        partial
+    }
+}
+
+/// Invert constraints on all partials in `bindings` and return them.
+fn invert_constraints(bindings: BindingStack) -> BindingStack {
+    let mut inverter = ConstraintInverter::new();
+    for Binding(_, value) in bindings.iter() {
+        inverter.fold_term(value.clone());
+    }
+    inverter.new_bindings
```

Comment:
In your example, `z` would be bound to `y` (a temp) in the final conjunct (`z = y`), so the binding it spits out would be `z => Variable(_y_8)`.

HOWEVER, a simplified version `f(x) if not (y = 1 and x.foo = y);` _did_ surface a bug. It was coming out as `partial(x) { [ partial(_temp) { [ _this != 1 ] } != _this.foo ] }` (two `!=` operations) because we were recursively inverting beyond the top-level constraints. We fixed that in 090bf02, and now the inner `_this = 1` partial simplifies to `1` and now we get `1 != _this.foo`.